### ISSUE TITLE
specialise and specialize

### ISSUE THREAD

**Issue by roboguy13** [ku-fpg/hermit#128]
There seem to be two commands, `specialise` and `specialize`. `specialise` is in `HERMIT.Dictionary.Rules` and `specialize` is in `HERMIT.Dictionary.Unfold`. They have slightly different documentation and I can't tell if they do they same thing or not. If they do, they should be merged into the same module and given the same documentation. Otherwise, one should be given a different name.

**Comment by ecaustin** [comment]
They do very different things.  `specialise` from `HERMIT.Dictionary.Rules` is a module-level rewrite that corresponds to GHC's specialization pass.  `specialize` from `HERMIT.Dictionary.Unfold` is an expression-level rewrite that specializes an application and inlines the result.

@xich should probably correct me if I'm wrong, but if `specialize` is a term-level rewrite only and it requires a beta-redex with a `Type` or `Coercion` argument then maybe something like `betaReduceTyCoPlusR` is a better name?  Or does that not properly indicate the subsequent inlining?

**Comment by xich** [comment]
Yeah, `specializeR` should be renamed... it performs an unfold only if all the arguments are types and coercions. This is somewhat useful only because type/coercion arguments tend to appear to the left of value arguments. So, for instance:

```
map :: forall a b. (a -> b) -> [a] -> [b]
```
The first two arguments to map are the types `a` and `b`. An application in Core will look like:

```
map Char Int digitToInt xs
```

Applying `onetdR specializeR`:

```
(\ f xs -> case xs of
                 [] -> []
                 (y:ys) -> f y : map Char Int f ys) digitToInt xs
```

(And to be totally honest, I think `specializeR` was defined mostly to demonstrate a use for `unfoldPredR`.)

**Comment by ecaustin** [comment]
I actually implemented an identical rewrite for my `haskhol-hermit` plugin because I didn't know `specializeR` existed (if it even did at that time), so it's definitely a useful transformation that serves a purpose.

I came up with the name `betaReduceTyCoPlusR` to try and mirror the `betaReducePlusR` rewrite that exists in the same module.  Are there any other related name pairings we should be following, e.g. would `tyCoBetaReducePlusR` be a more "faithful" HERMIT name?

**Comment by xich** [comment]
Hmmm, it's a hard one to name. We generally start the name with the 'verb' of the general operation. In this case, that would be either 'inline' or 'unfold'... the latter being inling+beta-reduction.

So I would probably go with `unfoldTyCoArgsR` or `unfoldTyCoAppR`.
